### PR TITLE
[2.7] bpo-31530: Stop crashes when iterating over a file on multiple threads.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue31530

### CODE REVIEW

File: Lib/test/test_file2k.py
Comment (vstinner) [inline Lib/test/test_file2k.py:655]: Please add a comment with a least "bpo-31530".

File: Lib/test/test_file2k.py
Comment (vstinner) [inline Lib/test/test_file2k.py:669]: Please add a comment with a least "bpo-31530".

File: Objects/fileobject.c
Comment (vstinner) [inline Objects/fileobject.c]: I would remove the final ".", it's uncommon in Python error messages.

**Review by vstinner — APPROVED** [review]
LGTM, but I would prefer to see reviews of other core developers before seeing this change merged. Especially because @benjaminp proposed PR #3670 with a different approach.

File: Misc/NEWS.d/next/Core and Builtins/2017-09-20-18-28-09.bpo-31530.CdLOM7.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2017-09-20-18-28-09.bpo-31530.CdLOM7.rst:1]: I suggest to complete the description:

seek() and next() methods of file objects now raise an exception during concurrent operation on the same file object. A lock can be used to prevent the error.

File: Objects/fileobject.c
Comment (serhiy-storchaka) [reply Objects/fileobject.c]: I agree. But this copies the error message at line 431. Do you want to remove the final "." from all error messages in this file?

**Review by serhiy-storchaka — COMMENTED** [review]
Yes, and I would want to see reviews of @benjaminp.

File: Objects/fileobject.c
Comment (vstinner) [reply Objects/fileobject.c]: Oh. I looked quickly at fileobject.c error messages, but I didn't notice that the error message already existed in close(). That's a good reason to generalize the check for concurrent operation :-)

In this case, it's fine, keep the final dot :-)

**Comment by vstinner** [comment]
Oh, you removed the final dot. I'm fine with that too. It's just a minor thing, I don't care much :-) (But I like the idea of being consistent in error messages.)

File: Lib/test/test_file2k.py
Comment (vstinner) [reply Lib/test/test_file2k.py:655]: done

File: Lib/test/test_file2k.py
Comment (vstinner) [reply Lib/test/test_file2k.py:669]: done

File: Misc/NEWS.d/next/Core and Builtins/2017-09-20-18-28-09.bpo-31530.CdLOM7.rst
Comment (vstinner) [reply Misc/NEWS.d/next/Core and Builtins/2017-09-20-18-28-09.bpo-31530.CdLOM7.rst:1]: done

**Review by vstinner — APPROVED** [review]
LGTM.

(But again, I would like to see a review of at least another core dev, especially @benjaminp.)

File: Lib/test/test_file2k.py
Comment (benjaminp) [inline Lib/test/test_file2k.py:666]: It's probably a better stress test to continue looping even when you get an exception.

File: Lib/test/test_file2k.py
Comment (benjaminp) [inline Lib/test/test_file2k.py:685]: similarly here

**Review by benjaminp — COMMENTED** [review]
The reason I wrote my patch is that it seems weird that concurrent f.read() works but not iteration. This is probably fine, though, since no one can have been relying on the current behavior.

`drop_readahead` could also use a `assert(f->unlocked_count == 0);` assert

File: Lib/test/test_file2k.py
Comment (serhiy-storchaka) [reply Lib/test/test_file2k.py:666]: This makes the test tooooooo slow. I have killed it after 8 minutes.

Comment (benjaminp) [reply Lib/test/test_file2k.py:666]: Decrease the size of the file then?

Comment (serhiy-storchaka) [reply Lib/test/test_file2k.py:666]: `2**12` lines is not enough for provoking a crash, `2**13` to `2**14` lines provoke random crashes, and `2**15` lines is needed for stable crashing. After patching the test with `2**13` lines takes about 40 sec, and `2**15` lines need 3 minutes to finish. This is still too long.

**Review by benjaminp — APPROVED** [review]
Okay. Please put an `assert` in `drop_readahead`, though.

**Comment by serhiy-storchaka** [comment]
> `drop_readahead` could also use a `assert(f->unlocked_count == 0);` assert

This looks unsafe. `f->unlocked_count` can be non-zero if other thread performs non-read operation like tell(). And more, it seems to me that the check should be moved inside `readahead_get_line_skip()` because it can be called recursively, after releasing/acquiring GIL that allows other thread to increase `f->unlocked_count`.

**Pull request merged**
